lightning/src/ln/inbound_payment.rs

```diff
@@ -63,6 +65,49 @@ impl ExpandedKey {
 			user_pmt_hash_key,
 		}
 	}
+
+	/// Returns an [`HmacEngine`] used to construct [`Offer::metadata`].
+	///
+	/// [`Offer::metadata`]: crate::offers::offer::Offer::metadata
+	pub(crate) fn hmac_for_offer(&self, nonce: Nonce) -> HmacEngine<Sha256> {
+		let mut hmac = HmacEngine::<Sha256>::new(&self.ldk_pmt_hash_key);
+		hmac.input(&nonce.0);
+		hmac
+	}
+
+	pub(crate) fn signing_pubkey_for_offer(&self, nonce: Nonce) -> PublicKey {
+		let secp_ctx = Secp256k1::new();
+		let mut tweak = [0; 32];
+		tweak[..Nonce::LENGTH].copy_from_slice(&nonce.0);
+
+		SecretKey::from_slice(&self.ldk_pmt_hash_key)
+			.unwrap()
+			.mul_tweak(&Scalar::from_be_bytes(tweak).unwrap())
+			.unwrap()
+			.public_key(&secp_ctx)
+	}
+}
+
+///
+#[derive(Clone, Copy)]
+pub struct Nonce(pub(crate) [u8; Self::LENGTH]);
+
+impl Nonce {
```

Comment:
It's kinda nice having a type even if it wraps a `u128` (though that adds a conversion when checking the metadata) as it encapsulates creation and properties like length and byte order. Otherwise, you'd have places in the code that need to get those right.